### PR TITLE
KAFKA-16040; Rename `Generic` to `Classic`

### PR DESCRIPTION
**Pull request by dajac**
People has raised concerned about using `Generic` as a name to designate the old rebalance protocol. We considered using `Legacy` but discarded it because there are still applications, such as Connect, using the old protocol. We settled on using `Classic` for the `Classic Rebalance Protocol`.

The changes in this patch are extremely mechanical. It basically replaces the occurrences of `generic` by `classic`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetrics.java [hunk 1]
```diff
@@ -319,40 +319,40 @@ public Long value() {
...
 
         registry.newGauge(NUM_GENERIC_GROUPS_DEAD, new com.yammer.metrics.core.Gauge<Long>() {
             @Override
             public Long value() {
-                return numGenericGroups(GenericGroupState.DEAD);
+                return numGenericGroups(ClassicGroupState.DEAD);
             }
         });
 
         registry.newGauge(NUM_GENERIC_GROUPS_EMPTY, new com.yammer.metrics.core.Gauge<Long>() {
```
Comment (divijvaidya) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetrics.java]: Are we keeping the metrics name with "generic" keyword though? I think we should if these metrics have already been released in previous versions, if not, might want to change the metric names to "classic" as well.

e.g. GENERIC_GROUP_REBALANCES_SENSOR_NAME

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetrics.java]: We don't have any existing metrics using "generic". They are all new in 3.7.

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetrics.java]: Regarding `GENERIC_GROUP_REBALANCES_SENSOR_NAME`, the patch only changes the name of the sensor. The metrics are not touched.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetrics.java [hunk 2]
```diff
@@ -70,11 +70,11 @@ public class GroupCoordinatorMetrics extends CoordinatorMetrics implements AutoC
     public static final String OFFSET_COMMITS_SENSOR_NAME = "OffsetCommits";
     public static final String OFFSET_EXPIRED_SENSOR_NAME = "OffsetExpired";
     public static final String OFFSET_DELETIONS_SENSOR_NAME = "OffsetDeletions";
-    public static final String GENERIC_GROUP_COMPLETED_REBALANCES_SENSOR_NAME = "CompletedRebalances";
-    public static final String GENERIC_GROUP_REBALANCES_SENSOR_NAME = "GenericGroupRebalances";
+    public static final String CLASSIC_GROUP_COMPLETED_REBALANCES_SENSOR_NAME = "CompletedRebalances";
+    public static final String CLASSIC_GROUP_REBALANCES_SENSOR_NAME = "GenericGroupRebalances";
```
Comment (divijvaidya) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetrics.java]: I should have added my previous comment over here. Let's continue conversation at this place.

I understand that we have only changed the variable name of the sensor here. My question is, should we change "GenericGroupRebalances" to "ClassicGroupRebalances" here if this has just been introduced in 3.7?

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetrics.java]: Indeed, we should rename this one. Missed it.

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/GroupCoordinatorMetrics.java]: Fixed.

**Comment by lucasbru** [comment]
Please check
* `GroupCoordinatorConfig.GENERIC_GROUP_NEW_MEMBER_JOIN_TIMEOUT_MS`
* Some comments in `GroupCoordinator`
* Various names in `GroupCoordinatorMetrics`
Are some of these remaing for backwards compatibility?

**Comment by dajac** [comment]
@lucasbru Thanks. I just fixed them.

**Review by lucasbru — APPROVED** [review]
LGTM! Thanks, @dajac

**Review by divijvaidya — APPROVED** [review]

**Comment by dajac** [comment]
As it is really hard to get a clean build due to OOM errors (also in trunk), I have verified the build locally. I am going to merge it to trunk and 3.7.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/LegacyKafkaConsumer.java [hunk 3]
```diff
@@ -96,13 +96,13 @@
...
 /**
- * A client that consumes records from a Kafka cluster using the {@link GroupProtocol#GENERIC generic group protocol}.
+ * A client that consumes records from a Kafka cluster using the {@link GroupProtocol#CLASSIC classic group protocol}.
  * In this implementation, all network I/O happens in the thread of the application making the call.
  *
  * <p/>
  *
  * <em>Note:</em> per its name, this implementation is left for backward compatibility purposes. The updated consumer
- * group protocol (from KIP-848) introduces allows users continue using the legacy "generic" group protocol.
+ * group protocol (from KIP-848) introduces allows users continue using the legacy "classic" group protocol.
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/LegacyKafkaConsumer.java:105]: there's a bit of grammatical confusion here, perhaps: `group protocol ... allows user continue...`

**Pull request merged**